src/app/file-upload-modal/file-upload-modal.component.ts

```diff
@@ -0,0 +1,225 @@
+import { Component, OnInit, ChangeDetectorRef, ElementRef, ViewChild, HostListener } from '@angular/core';
+import { FormControl } from '@angular/forms';
+import { NgxSmartModalService } from 'ngx-smart-modal';
+import { Utils } from 'app/shared/utils/utils';
+import { uniqBy } from 'lodash';
+import { UploadedFile } from './types';
+import { ModalData } from 'app/shared/types/modal';
+
+
+@Component({
+  selector: 'app-file-upload-modal',
+  templateUrl: './file-upload-modal.component.html',
+  styleUrls: ['./file-upload-modal.component.scss']
+})
+export class FileUploadModalComponent implements OnInit {
+  closeDialog: boolean;
+  modalData: ModalData;
+  uploadedFiles: UploadedFile[];
+  file: UploadedFile;
+  fileList: UploadedFile[];
+  fileExt: string;
+  selectedFiles: UploadedFile[];
+  @ViewChild("fileBrowser") fileBrowser: ElementRef;
+  @ViewChild("fileInfoForm") fileInfoForm: ElementRef;
+
+  constructor(
+    public utils: Utils,
+    private ngxSmartModalService: NgxSmartModalService,
+    private _changeDetectorRef: ChangeDetectorRef,
+  ) { }
+
+  ngOnInit() {
+    this.closeDialog = false;
+    this.selectedFiles = [];
+    this.fileList = [];
+    this.uploadedFiles = [];
+    this.modalData = this.ngxSmartModalService.getModalData('file-upload-modal');
+
+    if (this.modalData.fileExt) {
+      this.fileExt = this.modalData.fileExt;
+    } else {
+      this.fileExt = 'jpg, jpeg, gif, png, bmp, doc, docx, xls, xlsx, ppt, pptx, pdf, txt, zip';
+    }
+  }
+
+  /**
+   * If there are uploaded files, or files with updated info, show the close
+   * dialog to confirm the user wants to close. Otherwise, close modal.
+   *
+   * @returns {void}
+   */
+  handleClose(): void {
+    if (this.isSaveRequired()) {
+      this.closeDialog = true;
+    } else {
+      this.ngxSmartModalService.close('file-upload-modal');
+    }
+  }
+
+  /**
+   * Stop showing the close dialog.
+   *
+   * @returns {void}
+   */
+  handleCancelClose(): void {
+    this.closeDialog = false;
+  }
+
+  /**
+   * Adds a pseudo-random ID and alt tag fields to uploaded files.
+   * Then puts the new files at the beginning of the fileList.
+   *
+   * @param uploadedFiles The files uploaded by the file-upload component.
+   * @returns {void}
+   */
+  uploadFiles( uploadedFiles: UploadedFile[] ): void {
+    this.uploadedFiles = uploadedFiles.map(file => {
+      // Create a temporary pseudo-random ID to help with sorting.
+      file.id = Math.floor(Math.random() * 1000000000000);
+      file.alt = new FormControl('');
+      return file;
+    })
+
+    this.uploadedFiles.forEach(file => {
+      this.fileList.unshift(file);
+    });
+
+    this._changeDetectorRef.detectChanges();
+  }
+
+  /**
+   * Add file to selected files list.
+   * Checks that the number of files in the list does not exceed fileNum.
```

Comment:
just saying without context fileNum is not making a lot of sense, especially in the sense of modalData.fileNum.  Maybe have a comment where it checks so people know the reason for this check, or rename variable.  Maybe this isn't even your variable.. In this case, just let the developer know what the check is for on 103.

Reply:
Right, that's fair enough. In a future commit, I'll revise the comment to:
```suggestion
   * Checks that the number of files in the list does not exceed the maximum number of allowed selectable files.
```